### ISSUE TITLE
EADDRINUSE errors when tests run concurrently

### ISSUE THREAD

**Issue by roberttaylor426** [httptoolkit/mockttp#42]
We have multiple tests in our test suite that depend on mockttp, using its capability to randomly assign a port, and when Jest runs these tests concurrently we occasionally see errors in the form:

```bash
listen EADDRINUSE: address already in use :::8000
```

To reproduce this issue we set up the following test:

```typescript
import { getLocal } from 'mockttp';

describe('mockttp tests running in parallel', () => {
    const mockServer = getLocal();

    beforeAll(async () => {
        await mockServer.start();
    });

    afterAll(async () => await mockServer.stop());

    it('should pass', () => {
        expect(true).toBe(true);
    });
});
```

and then executed:

```bash
printf %s\\n {0..64} | xargs -n 1 -P 16 npx jest <test-name>
```

to run 16 instances of the test in parallel (on our 16 core machine) 4 times in a row. On most runs we see this issue occurring.

**Comment by pimterry** [comment]
> Can we introduce some kind of locking or retry mechanism in mockttp-server.start, or do you think we're using the library in a way it wasn't intended to be used?

Fortunately, it already has one!

https://github.com/httptoolkit/mockttp/blob/7f536f0ddf81f09569d723ef94ce33127afb1248/src/server/mockttp-server.ts#L160-L169

Unfortunately it turns out it hasn't actually been working for a while, it was broken by a bug when `portParam` changed to support specifying port ranges.

I've now fixed this, and added a proper test that'll catch this in future: https://github.com/httptoolkit/mockttp/commit/415dea51e6aea45edd6653f6cf9b132af7249019. That test reliably fails on 0.21.2, and now seems rock solid. I'll get that all released ASAP.

Good catch, thanks very much for reporting this!

**Comment by roberttaylor426** [comment]
Lovely, thanks! :+1:

**Comment by roberttaylor426** [comment]
Thanks, I've updated to `1.0.0` and this seems to have resolved the stability issue.

Seemingly because other dependencies have also updated `tsc` is now complaining with:

```shell
node_modules/mockttp/dist/client/mockttp-client.d.ts:11:15 - error TS2304: Cannot find name 'Response'.
11     response: Response;
                 ~~~~~~~~

node_modules/mockttp/dist/client/mockttp-client.d.ts:12:44 - error TS2304: Cannot find name 'Response'.
12     constructor(message: string, response: Response);
                                              ~~~~~~~~

node_modules/mockttp/dist/client/mockttp-client.d.ts:18:27 - error TS2304: Cannot find name 'Response'.
18     constructor(response: Response, errors: Array<{
                             ~~~~~~~~
```

I had to work around this by adding `dom` as a `lib` dependency in my `tsconfig.json`. Given this is a backend project this felt a little unfortunate; is there a better way?

Happy to raise a separate issue if needed.

**Comment by pimterry** [comment]
> I had to work around this by adding dom as a lib dependency in my tsconfig.json. Given this is a backend project this felt a little unfortunate; is there a better way?

Agreed, that shouldn't be required. I think this is only because you were explicitly setting a `lib` property before. It looks like `dom` is included in the default value of `lib`, which is why that hasn't come up elsewhere.

I'm not sure why this would've just started happening though, those types haven't changed since the previous version. Any ideas?

I think this shouldn't be too bad to fix anyway, we just need to bundle the fetch types, which are already available for node as part of `node-fetch` anyway (which is what this code is actually using, when running in node).

**Comment by roberttaylor426** [comment]
> I'm not sure why this would've just started happening though, those types haven't changed since the previous version. Any ideas?

Is a great question, I'm really not sure. In our code base the `tsc` error was introduced somewhere between `0.19.1` and `0.19.2` (we were previously on `0.19.0`). Judging from [the diff](https://github.com/httptoolkit/mockttp/compare/v0.19.1...v0.19.2) however it's not clear to me what might have caused things to break.

> I think this shouldn't be too bad to fix anyway, we just need to bundle the fetch types, which are already available for node as part of node-fetch anyway (which is what this code is actually using, when running in node).

This would be great :+1:

**Comment by pimterry** [comment]
Turns out including node-fetch and using those types is a bit more complicated than it sounds.

Instead, in https://github.com/httptoolkit/mockttp/commit/0dbce78fcc4c251ce4d614b244a5204af4cfdf67 I've gone for the current best [placeholder types](https://github.com/microsoft/TypeScript/issues/31894) workaround of forward declaring the global `Response` type. That seems to solve this cleanly enough, without including all the dom types unnecessarily, and should solve out your issue. I'll get that released tomorrow.

**Comment by roberttaylor426** [comment]
Great, thank you :+1: 

I look forward to giving this a try.

**Comment by pimterry** [comment]
This is now published as v1.0.1 :+1:
